### PR TITLE
[SPARK-27199][SQL] Replace TimeZone by ZoneId in TimestampFormatter API

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to use `ZoneId` instead of `TimeZone` in:
- the `apply` and `getFractionFormatter ` methods of the `TimestampFormatter` object,
- and in implementations of the `TimestampFormatter` trait like `FractionTimestampFormatter`.

The reason of the changes is to avoid unnecessary conversion from `TimeZone` to `ZoneId` because `ZoneId` is used in `TimestampFormatter` implementations internally, and the conversion is performed via `String` which is not for free. Also taking into account that `TimeZone` instances are converted from `String` in some cases, the worse case looks like `String` -> `TimeZone` -> `String` -> `ZoneId`. The PR eliminates the unneeded conversions.

## How was this patch tested?

It was tested by `DateExpressionsSuite`, `DateTimeUtilsSuite` and `TimestampFormatterSuite`.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala [hunk 1]
```diff
@@ -43,18 +43,17 @@ sealed trait TimestampFormatter extends Serializable {
 }
 
 class Iso8601TimestampFormatter(
-    pattern: String,
-    timeZone: TimeZone,
-    locale: Locale) extends TimestampFormatter with DateTimeFormatterHelper {
+     pattern: String,
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala]: nit: indent

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala]: thanx. Just wondering why `scalastyle` didn't catch that.

**Comment by srowen** [comment]
Looks reasonable to me. Is there any behavior change? I don't think so, just checking.

**Comment by MaxGekk** [comment]
>  Is there any behavior change?

Potentially we could observe a behavoir change in parsing time zone id strings but because we did it twice in `TimeZone` and `ZoneId`, and now only in `ZoneId`. This is not the case.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**